### PR TITLE
Update box model to border box

### LINKED ISSUE #31137: [Issue] Update box model to border box
This issue is automatically created based on existing pull request: magento/magento2#31127: Update box model to border box

---------

### Description (*)

Why need to update this ?
Magento have define .lib-set-default-border-box but it's never used in code base
When use box-sizing border box we make frontend dev life easier. Only define in one file and apply to all other place. No need re-define box-sizing border box anymore
Now a lot of web use border-box model as reset default

When border-box is used padding-left and padding-right are not included in the calculation box width

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Questions or comments

Need workaround check visual both luma/blank theme when apply border-box model
In case this PR approve/accept i will check further make sure everything work like before no visual change

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by mrtuvn**
### Description (*)

Why we need update
box-sizing supported by all major browsers. Later allow make and craft markup grid easier

We switch the global box-sizing value to border-box. This ensures padding does not affect the final computed width of an element

#1 Update box-model to border-box 
Magento have define .lib-set-default-border-box but it's never used in code base
When use box-sizing border box we make frontend dev life easier. Only define in one file and styles apply to all other place for all element.
we can use content-box or padding-box (where supported) as well, without worrying about a universal selector overriding css

box-sizing border box defined in file app/design/frontend/Magento/blank/web/css/source/_reset.less

Screen Shot 2020-12-03 at 10 38 15

https://thisthat.dev/border-box-vs-content-box
Now a lot of web use border-box model as default

When border-box is used border,  padding-left and padding-right are not included in the calculation box width

all elements can re-define box model (content-box) without any issue if need. Default all elements in html will have border box model

This PR only remove define border-box in theme core/libs magento in luma theme
Keep old declare mixins related boxder-box for maintain compatible in 3rd-party themes
Keep backward compatible in blank theme (avoid too much changes in blank)

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. Check all pages blank/luma theme make sure all page work without break, glitch, overlap box

### Questions or comments

Old article related
https://css-tricks.com/inheriting-box-sizing-probably-slightly-better-best-practice
Need workaround check visual both luma/blank theme when apply border-box model
Should test all screen make sure everything work like before no big visual change break any elements

CC: @krzksz @Leland @ptylek  good to hear feedback from you guys
@DanielRuf ,@Parakoopa, @kirmorozov ,@leoquijano, @jonathanKingston can you guys give me some opinion

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#31137: Update box model to border box

### CODE REVIEW

**Comment by sidolov** [comment]
@magento create issue

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento give me 2.4-develop instance

**Comment by mrtuvn** [comment]
@magento give me test instance

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

**Comment by mrtuvn** [comment]
@krzksz @ihor-sviziev can we have priority review and approve this soon ?

**Comment by mrtuvn** [comment]
@magento give me test instance

**Comment by ihor-sviziev** [comment]
@mrtuvn this looks like a big change that might cause some issues during the update. I believe it should be delivered in 2.5

File: app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml [hunk 1]
```diff
@@ -7,7 +7,7 @@
 -->
 <page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
     <head>
-        <meta name="viewport" content="width=device-width, initial-scale=1"/>
+        <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1.0, user-scalable=no"/>
```
Comment (ihor-sviziev) [inline app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml]: could you explain why this change is needed?

File: app/design/frontend/Magento/blank/Magento_Catalog/web/css/source/module/_listings.less [hunk 2]
```diff
@@ -34,7 +34,7 @@
         }
 
         &-item {
-            font-size: 1.4rem;
+            .lib-font-size(14);
```
Comment (ihor-sviziev) [inline app/design/frontend/Magento/blank/Magento_Catalog/web/css/source/module/_listings.less]: Could you explain why it's needed?

File: app/design/frontend/Magento/luma/Magento_Catalog/web/css/source/module/_toolbar.less [hunk 3]
```diff
@@ -170,7 +170,6 @@
     }
 
     .modes {
-        display: inline-block;
```
Comment (ihor-sviziev) [inline app/design/frontend/Magento/luma/Magento_Catalog/web/css/source/module/_toolbar.less]: could you explain why it was removed?

File: lib/web/css/source/lib/_sections.less [hunk 4]
```diff
@@ -17,7 +17,6 @@
         width: auto;
 
         > .switch {
-            .lib-css(height, @_tab-control-height__base);
```
Comment (ihor-sviziev) [inline lib/web/css/source/lib/_sections.less:20]: could you explain why it was removed ?

File: lib/web/css/source/lib/_sections.less [hunk 5]
```diff
@@ -106,7 +104,6 @@
             .lib-css(background, @_tab-control-background-color);
             .lib-css(border, @_tab-content-border);
             border-bottom: none;
-            .lib-css(height, @_tab-control-height);
```
Comment (ihor-sviziev) [inline lib/web/css/source/lib/_sections.less:109]: and this one?

File: lib/web/css/source/lib/_sections.less [hunk 6]
```diff
@@ -244,7 +239,6 @@
             .lib-css(border-left,@_accordion-control-border-left);
             .lib-css(border-right,@_accordion-control-border-right);
             .lib-css(border-top,@_accordion-control-border-top);
-            .lib-css(height, @_accordion-control-height);
```
Comment (ihor-sviziev) [inline lib/web/css/source/lib/_sections.less:247]: and this

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by mrtuvn** [comment]
Yes i think we should improve that. Seem a lot of things will change. Can it will be a part of codebase in early realease

File: app/design/frontend/Magento/blank/Magento_Catalog/web/css/source/module/_listings.less (hunk 2 above)
Comment (mrtuvn) [reply app/design/frontend/Magento/blank/Magento_Catalog/web/css/source/module/_listings.less]: This is standard font mixins default of magento. It's will work same as default

File: app/design/frontend/Magento/luma/Magento_Catalog/web/css/source/module/_toolbar.less (hunk 3 above)
Comment (mrtuvn) [reply app/design/frontend/Magento/luma/Magento_Catalog/web/css/source/module/_toolbar.less]: When use float element will treated as block element so basically display define will unnecessary

File: app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml (hunk 1 above)
Comment (mrtuvn) [reply app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml]: As I mentioned in description 
This update meta viewport content to ensure proper rendering and touch zooming for all devices

File: lib/web/css/source/lib/_sections.less (hunk 4 above)
Comment (mrtuvn) [reply lib/web/css/source/lib/_sections.less:20]: No need fixed height here

File: lib/web/css/source/lib/_sections.less (hunk 5 above)
Comment (mrtuvn) [reply lib/web/css/source/lib/_sections.less:109]: same as above

File: lib/web/css/source/lib/_sections.less (hunk 6 above)
Comment (mrtuvn) [reply lib/web/css/source/lib/_sections.less:247]: same as above

File: app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml (hunk 1 above)
Comment (mrtuvn) [reply app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml]: i think should revert this back to origin as this change is unneccessary. Depend on projects  require we need to modify viewport in custom theme

File: lib/web/css/source/lib/_sections.less (hunk 4 above)
Comment (mrtuvn) [reply lib/web/css/source/lib/_sections.less:20]: Screenshot from 2020-12-11 22-26-18

**Comment by ihor-sviziev** [comment]
@mrtuvn I'll check it. Just I'm curious that such a change might break the websites during the patch Magento update.

**Comment by mrtuvn** [comment]
From my check not much change compare with previous version(use default box-model content-box). This PR changes will apply for frontend only. Affected for both luma and blank

**Comment by mrtuvn** [comment]
@magento run all tests

File: lib/web/css/source/lib/_sections.less (hunk 4 above)
Comment (mrtuvn) [reply lib/web/css/source/lib/_sections.less:20]: And this tab in theme blank
Screenshot from 2020-12-13 11-27-02

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@mrtuvn ok. Let's try to deliver it in 2.4 release line. Removed 2.5 milestone

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE, Functional Tests EE

**Comment by mrtuvn** [comment]
Seem fails functional tests come from another modules

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento give me test instance
CC: @ihor-sviziev

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run all tests

File: app/design/frontend/Magento/luma/Magento_Catalog/web/css/source/module/_toolbar.less (hunk 3 above)
Comment (mrtuvn) [reply app/design/frontend/Magento/luma/Magento_Catalog/web/css/source/module/_toolbar.less]: Change here previously remove line will cause missing modes switch toolbar in desktop screen
I have updated that

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8552](https://jira.corp.magento.com/browse/ENGCOM-8552) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by mrtuvn** [comment]
Can we have QA verify for this PR

**Comment by ihor-sviziev** [comment]
@engcom-Alfa is there any progrss?

**Comment by mrtuvn** [comment]
rebased code
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests
In the mean times no tests available so i rebase + add some clean up code

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests EE, WebAPI Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests

**Comment by mrtuvn** [comment]
@magento give me 2.4-develop instance

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests CE, Functional Tests B2B, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests CE

**Comment by mrtuvn** [comment]
@magento run all tests

File: app/design/frontend/Magento/blank/web/css/source/_reset.less [hunk 7]
```diff
@@ -4,5 +4,7 @@
 //  */
 
 & when (@media-common = true) {
+    .lib-html5doctor-reset();
```
Comment (Leland) [inline app/design/frontend/Magento/blank/web/css/source/_reset.less]: This feels to me like we're including a lot of properties that we probably don't actually need. What aspects of this reset are not already being met by the generic `lib-magento-reset` mixin below it?

This line for instance: 

https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/lib/web/css/source/lib/_resets.less#L439-L442

Is mostly a dupe of this block, from `lib-magento-reset`:

https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/lib/web/css/source/lib/_resets.less#L20-L32

Perhaps rather than calling this mixin directly, add the needed lines from the html5doctor mixin to the generic reset mixin instead?

Comment (mrtuvn) [reply app/design/frontend/Magento/blank/web/css/source/_reset.less]: Thank you for your comment!! @Leland 
As frontend view can you give me more review on this PR

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run WebAPI Tests, Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@mrtuvn,
I still believe this is a too high risk of breaking frontend, so you don't expect having it in a small release, and we should post stone it to 2.5.

**Comment by mrtuvn** [comment]
yes totally agreed. I don't think this one can be accept by internal even in 2.5. Since this one can possible introduce unexpected broken layout
